src/lib/core/CHIPError.h

```diff
@@ -102,14 +102,15 @@ class ChipError
      */
     enum class SdkPart : uint8_t
     {
-        kCore            = 0, ///< SDK core errors.
-        kInet            = 1, ///< Inet layer errors; see <inet/InetError.h>.
-        kDevice          = 2, ///< Device layer errors; see <platform/CHIPDeviceError.h>.
-        kASN1            = 3, ///< ASN1 errors; see <asn1/ASN1Error.h>.
-        kBLE             = 4, ///< BLE layer errors; see <ble/BleError.h>.
-        kIMGlobalStatus  = 5, ///< Interaction Model global status code.
-        kIMClusterStatus = 6, ///< Interaction Model cluster-specific status code.
-        kApplication     = 7, ///< Application-defined errors; see CHIP_APPLICATION_ERROR
+        kCore                   = 0, ///< SDK core errors.
+        kInet                   = 1, ///< Inet layer errors; see <inet/InetError.h>.
+        kDevice                 = 2, ///< Device layer errors; see <platform/CHIPDeviceError.h>.
+        kASN1                   = 3, ///< ASN1 errors; see <asn1/ASN1Error.h>.
+        kBLE                    = 4, ///< BLE layer errors; see <ble/BleError.h>.
+        kIMGlobalStatus         = 5, ///< Interaction Model global status code.
+        kIMClusterStatusFailure = 6, ///< Interaction Model cluster-specific status code (failure)
+        kApplication            = 7, ///< Application-defined errors; see CHIP_APPLICATION_ERROR
+        kIMClusterStatusSuccess = 8, ///< Interaction Model cluster-specific status code (success)
```

Comment:
We sort of purposefully did not do this, because encoding a "success" status as a CHIP_ERROR is really weird and brittle.  For cases when we might have a cluster-specific success we _really_ need to be using a different return type than CHIP_ERROR, or have an out-param for the status or _something_.

Reply:
The difficulty was that conceptually Read/Write/Invoke are supposed to be StatusIB (so ClusterStatus) however for chunking we are also expecting a CHIP_ERROR to encode the out of space errors.

I am concerned that the ripple effect of that. The existing code logic is not very clean on single-return because `AddStatus` is separated from `error code return` and that allowed our types to diverge.

Reply:
One thing we could do:

1. Add an `mOutOfSpace` member to `AttributeValueEncoder`.
2. Set this member as needed in the relevant public `Encode` methods (we can do this without too much codesize hit, maybe).
3. Have whatever created the AttributeValueEncoder check that state (like it already checks the `mTriedEncode` state) and handle it appropriately.
4. Change `AttributeAccessInterface::Read` to return `ClusterStatus`
5. Change `DataModel::Provider::ReadAttribute` to return `ClusterStatus`

That does mean not being able to use ReturnErrorOnFailure() all over in `Read`, though....

Another option is to just hang the ClusterStatus off the AttributeValueEncoder and the few consumers that want a success status can do that, and we check for it in the one chokepoint at the callsite.